Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/mesh_to_point_set_hausdorff_distance.h

```diff
@@ -0,0 +1,362 @@
+// Copyright (c) 2016 INRIA Sophia-Antipolis (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+// You can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// Licensees holding a valid commercial license may use this file in
+// accordance with the commercial license agreement provided with the software.
+//
+// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+//
+// $URL$
+// $Id$
+//
+//
+// Author(s)     : Simon Giraudot and Maxime Gimeno
+
+#ifndef CGAL_MESH_TO_POINT_SET_HAUSDORFF_DISTANCE_H
+#define CGAL_MESH_TO_POINT_SET_HAUSDORFF_DISTANCE_H
+
+#include <CGAL/Orthogonal_k_neighbor_search.h>
+#include <CGAL/Search_traits_3.h>
+#include <CGAL/squared_distance_3.h>
+
+template <typename Kernel>
+class CPointH
+{
+private:
+
+  typedef typename Kernel::Point_3 Point;
+  typedef typename Kernel::FT FT;
+  Point m_point;
+  FT m_hausdorff;
+
+public:
+
+  CPointH ()
+  {
+    m_point = Point (0., 0., 0.);
+    m_hausdorff = 0.;
+  }
+  CPointH (const Point& p, FT h = 0.)
+  {
+    m_point = p;
+    m_hausdorff = h;
+  }
+  CPointH (const CPointH& p)
+  {
+    m_point = p ();
+    m_hausdorff = p.hausdorff ();
+  }
+
+  Point operator() () const { return m_point; }
+  Point& operator() () { return m_point; }
+  FT hausdorff () const { return m_hausdorff; }
+
+  static Point mid_point (const CPointH& a, const CPointH& b)
+  {
+    return Point (0.5 * (a().x() + b().x()),
+                  0.5 * (a().y() + b().y()),
+                  0.5 * (a().z() + b().z()));
+  }
+
+
+};
+
+template <typename Kernel>
+class CRefTriangle
+{
+
+private:
+
+  typedef typename Kernel::Point_3 Point;
+  typedef typename Kernel::FT FT;
+  typedef CPointH<Kernel> PointH;
+  PointH m_point[3];
+  int m_edge;
+  FT m_upper_bound;
+  FT m_lower_bound;
+  Point m_bisector;
+
+public:
+
+  CRefTriangle (const PointH& a, const PointH& b, const PointH& c) // Triangle
+  {
+    m_point[0] = a;
+    m_point[1] = b;
+    m_point[2] = c;
+
+    m_edge = -1;
+    for (unsigned int i = 0; i < 3; ++ i)
+    {
+      if (CGAL::angle (m_point[(i+1)%3](), m_point[i](), m_point[(i+2)%3]())
```

Comment:
@sgiraudot  note that this is not a robust predicate. No idea if it is critical.